ast/compile.go

```diff
@@ -2195,56 +2194,86 @@ func (c *Compiler) rewriteLocalVars() {
 				c.err(err)
 			}
 
-			// Rewrite assignments in body.
-			used := NewVarSet()
+			argsStack := newLocalDeclaredVars()
 
-			last := rule.Head.Ref()[len(rule.Head.Ref())-1]
-			used.Update(last.Vars())
+			args := NewVarVisitor()
+			args.Walk(rule.Head.Args)
+			unusedArgs := args.Vars()
 
-			if rule.Head.Key != nil {
-				used.Update(rule.Head.Key.Vars())
-			}
+			c.rewriteLocalArgVars(gen, argsStack, rule)
 
-			if rule.Head.Value != nil {
-				used.Update(rule.Head.Value.Vars())
-			}
+			for rule := rule; rule != nil; rule = rule.Else {
```

Comment:
It would be helpful to add some detailed comments to understand what is going on here. Maybe also move it in a separate function.

Are there any performance impacts from this change?

Reply:
I'll add some tactical documentation 👍 .

We do an extra walk and iteration over the args of a function. I considered putting that behind a `Compiler.strict` check, but I think the impact is so negligible I opted to keep the code cleaner instead.